### PR TITLE
Add fallback local url to marketo

### LINKED ISSUE #327: Firefox form bug
The form does not show up on this page if Enhanced tracking is enabled on Firefox for Windows OS on this page: https://goteleport.com/enterprise-trial/

image

### PR DESCRIPTION
**Pull request by iAdramelk**
Fix #327

Marketo didn't work in Firefox and Safari because of tracking cookie protection. This PR adds optional proxying for the Marketo request view goteleport.com domain if the original script is blocked.

To make it work we need to add following rewrites to nginx:

- `^js/forms2/(.*)` -> `http://app-ab44.marketo.com/js/forms2/$1`
- `^form/getForm` -> `http://app-ab44.marketo.com/form/getForm`
- `^form/getKnownLead` -> `http://app-ab44.marketo.com/form/getKnownLead`
- `^index.php/leadCapture/save2` -> `http://app-ab44.marketo.com/index.php/leadCapture/save2`

CC @klizhentas ^

### CODE REVIEW

**Comment by iAdramelk** [comment]
For details see https://blog.darwinapps.com/solution-to-embedded-marketo-forms-not-showing-up-in-firefox-c3780e6f8609

File: utils/marketo.ts [hunk 1]
```diff
@@ -28,21 +34,25 @@ interface UseMarketoProps {
...
 
     window.MktoForms2.loadForm(
-      NEXT_PUBLIC_MARKTO_BASE_URL,
+      url,
       NEXT_PUBLIC_MARKTO_MUNCHKIN_ID,
       formId,
       callback
     );
-  }, [loaded, formId, callback]);
+  }, [url, formId, callback]);
```
Comment (alex-kovoy) [inline utils/marketo.ts:55]: I would remove the `callback` from the dep list as changed callback is something that regularly happens. So it should not trigger multiple `window.MktoForms2.loadForm` in this case.

If you want to guarantee that this callback is always captured by that component, you can do this:

```ts
const ref = useRef(callback);

useEffect(() => {
    if (!url) {
      loadScript(NEXT_PUBLIC_MARKTO_BASE_URL)
        .catch(() => loadScript(NEXT_PUBLIC_HOST))
        .then(setUrl)
        .catch((e) => console.error(e));

      return;
    }

    const cb = () => ref.current();

    window.MktoForms2.loadForm(
      url,
      NEXT_PUBLIC_MARKTO_MUNCHKIN_ID,
      formId,
      cb);
}, [url, formId, ref]);      
```

**Review by alex-kovoy — APPROVED** [review]
with comments

**Review by klizhentas — CHANGES_REQUESTED** [review]
@iAdramelk this solution seems fragile - the moment Marketo switches the API, it would break and we would never know. What are the alternatives?

**Comment by iAdramelk** [comment]
> @iAdramelk this solution seems fragile - the moment Marketo switches the API, it would break and we would never know. What are the alternatives?

@klizhentas we may create separate domain/subdomain and redirect 100% calls from it to Marketo (or use cloudfire subdomain domain for this). This way even if Marketo API changes in the future, calls will still be made.

**Comment by klizhentas** [comment]
@iAdramelk I chatted to @alexwolfe and @kontsevoy. We want to switch our existing forms to Next engine and have a sync with marketo on the backend. Alex will let you know where is this on the priority list. Form requirements from me:

* Recaptcha just like on cloud form
* Marketo call on the backend with alert if it failed (and ideally retry)
* Consult with @alex-kovoy regarding security
* 
Please factor this out in a separate issue, and ask @alexwolfe about priority of this. You would need a list of forms from Alex with a spec and designs.

**Pull request closed without merging**